contracts/Uniswap/UniswapAnchoredView.sol

```diff
@@ -228,103 +198,66 @@ contract UniswapAnchoredView is AggregatorValidatorInterface, UniswapConfig, Own
     }
 
     /**
-     * @dev Fetches the current token/eth price accumulator from uniswap.
+     * @dev Fetches the latest TWAP from the UniV3 pool tick, over the last anchor period.
      */
-    function currentCumulativePrice(TokenConfig memory config) internal view returns (uint) {
-        (uint cumulativePrice0, uint cumulativePrice1,) = UniswapV2OracleLibrary.currentCumulativePrices(config.uniswapMarket);
+    function getUniswapTwap(TokenConfig memory config) internal view returns (uint256) {
+        uint32 anchorPeriod_ = uint32(anchorPeriod); // overflow is expected
+        uint32[] memory secondsAgos = new uint32[](2);
+        secondsAgos[0] = anchorPeriod_;
+        secondsAgos[1] = 0;
+        (int56[] memory tickCumulatives, ) = IUniswapV3Pool(config.uniswapMarket).observe(secondsAgos);
+
+        int24 timeWeightedAverageTick = int24((tickCumulatives[1] - tickCumulatives[0]) / anchorPeriod_);
+        uint160 sqrtPriceX96 = TickMath.getSqrtRatioAtTick(timeWeightedAverageTick);
+        // Squaring the result also squares the Q96 scalar (2**96),
+        // so after this mulDiv, the resulting TWAP is still in Q96 fixed precision.
+        uint256 twapX96 = FullMath.mulDiv(sqrtPriceX96, sqrtPriceX96, FixedPoint96.Q96);
+
+        uint256 twap;
         if (config.isUniswapReversed) {
-            return cumulativePrice1;
+            // In this case, we want the price of token1 relative to token0 instead,
+            // so calculate the inverse of the TWAP with a common precision (expScale).
+            // e.g. We want the ETH/USDC price but the actual pool is USDC/ETH.
+            // The TWAP is also down-scaled from Q96 in this step.
+            // -> twap = expScale / (twapX96 / (2**96))
+            twap = mul(expScale, 2**96) / twapX96;
```

Comment:
It's also possible to avoid this arithmetic by taking the negative of the cumulative tick because
<img src="http://www.sciweavers.org/tex2img.php?eq=%5Cbegin%7Baligned%7D%0Ap%20%26%3D%201.0001%5E%7Bt%7D%20%5C%5C%0A%28p%29%5E%7B-1%7D%20%26%3D%20%281.0001%5E%7Bt%7D%29%5E%7B-1%7D%20%5C%5C%0A%5Cfrac%7B1%7D%7Bp%7D%20%26%3D%201.0001%5E%7B-t%7D%0A%5Cend%7Baligned%7D&bc=White&fc=Black&im=jpg&fs=18&ff=mathdesign&edit=0" align="center" border="0" alt="\begin{aligned}p &= 1.0001^{t} \\(p)^{-1} &= (1.0001^{t})^{-1} \\\frac{1}{p} &= 1.0001^{-t}\end{aligned}" width="208" height="133" />

and do this instead

```solidity
if (config.isUniswapReversed) {
    timeWeightedAverageTick = -timeWeightedAverageTick;
}
```

Reply:
This division is dangerous. Could we use SafeMath.div() instead?

Reply:
Definitely in favour of using SafeMath ops - I had taken the approach of modifying as little as possible from the previous code (if not for readability in diffing).